ext/standard/proc_open.c

```diff
@@ -1203,38 +1248,46 @@ PHP_FUNCTION(proc_open)
 
 		close_descriptor(descriptors[i].childend);
 
-		if (descriptors[i].is_pipe) {
-			switch (descriptors[i].mode_flags) {
+		if (descriptors[i].type != DESCRIPTOR_TYPE_STD) {
+			if (descriptors[i].type == DESCRIPTOR_TYPE_PIPE) {
+				switch (descriptors[i].mode_flags) {
 #ifdef PHP_WIN32
-				case O_WRONLY|O_BINARY:
-					mode_string = "wb";
-					break;
-				case O_RDONLY|O_BINARY:
-					mode_string = "rb";
-					break;
+					case O_WRONLY|O_BINARY:
+						mode_string = "wb";
+						break;
+					case O_RDONLY|O_BINARY:
+						mode_string = "rb";
+						break;
 #endif
-				case O_WRONLY:
-					mode_string = "w";
-					break;
-				case O_RDONLY:
-					mode_string = "r";
-					break;
-				case O_RDWR:
-					mode_string = "r+";
-					break;
+					case O_WRONLY:
+						mode_string = "w";
+						break;
+					case O_RDONLY:
+						mode_string = "r";
+						break;
+					case O_RDWR:
+						mode_string = "r+";
+						break;
+				}
 			}
+
+			if (descriptors[i].type == DESCRIPTOR_TYPE_SOCKET) {
+				stream = php_stream_sock_open_from_socket((php_socket_t) descriptors[i].parentend, NULL);
+			} else {
 #ifdef PHP_WIN32
-			stream = php_stream_fopen_from_fd(_open_osfhandle((zend_intptr_t)descriptors[i].parentend,
-						descriptors[i].mode_flags), mode_string, NULL);
-			php_stream_set_option(stream, PHP_STREAM_OPTION_PIPE_BLOCKING, blocking_pipes, NULL);
+				stream = php_stream_fopen_from_fd(_open_osfhandle((zend_intptr_t)descriptors[i].parentend,
+							descriptors[i].mode_flags), mode_string, NULL);
+				php_stream_set_option(stream, PHP_STREAM_OPTION_PIPE_BLOCKING, blocking_pipes, NULL);
 #else
-			stream = php_stream_fopen_from_fd(descriptors[i].parentend, mode_string, NULL);
+				stream = php_stream_fopen_from_fd(descriptors[i].parentend, mode_string, NULL);
 # if defined(F_SETFD) && defined(FD_CLOEXEC)
-			/* Mark the descriptor close-on-exec, so it won't be inherited by
-			 * potential other children */
-			fcntl(descriptors[i].parentend, F_SETFD, FD_CLOEXEC);
+				/* Mark the descriptor close-on-exec, so it won't be inherited by
+				 * potential other children */
+				fcntl(descriptors[i].parentend, F_SETFD, FD_CLOEXEC);
```

Comment:
We have this code in two different places now for the socket and the pipe case. Can we keep it just here (outside the if/else)? (I'd also move the corresponding windows dup_handle code next to it, assuming I'm not missing some reason why these happen in different places.)